proposal.md

```diff
@@ -1,136 +1,239 @@
 # Immutable Collections
 
-* **Type**: Standard/kotlinx Library API proposal
-* **Author**: Ilya Gorbunov
+* **Type**: kotlinx library API proposal
+* **Author**: Ilya Gorbunov, Abduqodiri Qurbonzoda 
 * **Status**: Submitted
 * **Prototype**: In progress
 * **Discussion**: In the [issues](https://github.com/Kotlin/kotlinx.collections.immutable/issues) of this repository
 
 ## Summary
 
-Provide interfaces to represent truly immutable collections and maps.
+There are two families of collection interfaces in the Standard Library:
+
+* read-only interfaces, which only allow to read elements from a collection: 
+`Collection`, `List`, `Set`, `Map`;
+* mutable interfaces, which additionally allow to modify elements in a collection: 
+`MutableCollection`, `MutableList`, `MutableSet`, `MutableMap`.
+
+Kotlin mostly doesn't provide its own collection implementations: in Kotlin/JVM
+the collection interfaces are mapped to the standard JDK collection interfaces
+and implemented by the JDK collection implementations, such as `ArrayList`, `HashSet`, `HashMap` and other
+classes from `java.util` package.
+
+It is proposed to provide two families of interfaces that extend read-only collection interfaces:
+
+- Immutable collections: `ImmutableList`, `ImmutableSet`, etc. They specify by their contract the real immutability of their implementors.
+
+- Persistent collections: `PersistentList`, `PersistentSet`, etc. They extend immutable collections and provide efficient
+modification operations that return new instances of persistent collections with the modification applied. 
+The returned collections share parts of data structure with the original persistent collections.
+
+For persistent collection interfaces there shall be provided the following implementations:
+
+- default persistent list
+- default persistent hash set preserving the insertion order
```

Comment:
I am not sure if hash-based implementation detail should be exposed.
But, it's not a documentation, it's a proposal, so maybe it's ok.

Reply:
Using object hash to store/retrieve object or key is an important aspect of the implementation. It limits what objects can be safely stored in these collections and retrieved back (those that have their hash unchanged after placing them in a map/set).